### PR TITLE
Exit 1 when there are failures and add gateway4 backend in test cases

### PR DESCRIPTION
**Pull request by adelez**
Fixed a minor issue conversion between c++ and cxx.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_trickle.BM_PumpStreamServerToClient_Trickle_2M_32k.opt.new: 1


[trickle] No significant performance differences
```

**Comment by yongni** [comment]
If auth test works on GCE only, shall we make the test case generation more intelligent so we can still run the whole testcase file manually from local?

For example, instead of 

docker run --test_case=non-authtest
docker run --test_case=auth-test

Can we do
docker run ... --test_case=non-authtest
test <is_on_gce> && docker run ... --test_case=authtest
?

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by adelez** [comment]
There's definitely ways to work around this. Same for cloud gateways and gateways. But I'd like to do that in a separate PR, if you don't mind. I added comments.

**Comment by yongni** [comment]
Supporting "master" is a quick way to verify the sanity of the framework itself without relying on complex "git checkout" logic.  Without it, how do you plan to quickly test changes of the driver?

Also if we add auth tests like this, we lose the ability to run the test case script locally.  Can we at least add an env to skip auth tests during runtime so we can still run the generated test cases with command like 

docker_image=gcr.io/grpc-testing/grpc_interop_go1.7:master **skip_auth=1** ./testcases/go__master

?

**Comment by adelez** [comment]
When you create new tests, you don't need to change or add new images. Note that I'm not changing the version of the test cases, which is still just "master" right now. I'm just removing "master" images.

If we want to test the framework itself, I think it's better to add a test script. The master images are confusing as we don't know easily which commit it was created from. We probably should cleanup the GCS as well to remove all images that were created for testing the framework and only keep the official ones.

**Comment by adelez** [comment]
Reverted the change of removing master images. PTAL.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: tools/interop_matrix/create_testcases.sh [hunk 1]
```diff
@@ -52,11 +55,18 @@ function cleanup {
   [ -e "$CMDS_SH" ] && rm $CMDS_SH
 }
 trap cleanup EXIT
+# TODO(adelez): add test auth tests but do not run if not testing on GCE.
 # Running the testcases as sanity unless we are asked to skip.
 [ -z "$SKIP_TEST" ] && (echo "Running test cases: $CMDS_SH"; sh $CMDS_SH)
 
+# Convert c++ to cxx. 
+client_lang=$LANG
+if [ $LANG == "c++" ]
```
Comment (yongni) [inline tools/interop_matrix/create_testcases.sh]: nit (if-then style consistency with the rest of the code and avoid repeated assignment) :
  
if [$LANG == "c++" ]; then
  client_lang="cxx"
else
  client_lang=$LANG
fi

File: tools/interop_matrix/run_interop_matrix_tests.py [hunk 2]
```diff
@@ -180,9 +185,15 @@ def cleanup():
...
   docker_images = find_all_images_for_lang(lang)
   for runtime in sorted(docker_images.keys()):
-    run_tests_for_lang(lang, runtime, docker_images[runtime])
+    total_num_failures += run_tests_for_lang(lang, runtime, docker_images[runtime])
 
 report_utils.create_xml_report_file(_xml_report_tree, args.report_file)
+
+if total_num_failures:
+  sys.exit(1)
+else:
```
Comment (yongni) [inline tools/interop_matrix/run_interop_matrix_tests.py]: "else: " part is unnecessary

File: tools/interop_matrix/run_interop_matrix_tests.py [hunk 3]
```diff
@@ -48,9 +48,8 @@
 argp.add_argument('--gcr_path',
                   default='gcr.io/grpc-testing',
                   help='Path of docker images in Google Container Registry')
-
 argp.add_argument('--release',
-                  default='master',
+                  default='all',
```
Comment (yongni) [inline tools/interop_matrix/run_interop_matrix_tests.py]: now that you have changed the default, please update the instructions of README accordingly.

Also we have create_matrix_images.py which now has different default ('master') from this run_..._test.py.

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.counters.new: 1


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: tools/interop_matrix/create_testcases.sh (hunk 1 above)
Comment (adelez) [reply tools/interop_matrix/create_testcases.sh]: Done.

File: tools/interop_matrix/run_interop_matrix_tests.py (hunk 3 above)
Comment (adelez) [reply tools/interop_matrix/run_interop_matrix_tests.py]: Done.

File: tools/interop_matrix/run_interop_matrix_tests.py (hunk 2 above)
Comment (adelez) [reply tools/interop_matrix/run_interop_matrix_tests.py]: Done.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.old: 1


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by adelez** [comment]
Artifact Build Windows (internal CI)  #12911
Asan C #12912
Basic Tests C Linux [dbg] #12913
Basic Tests C++ Linux [dbg] #12873, #11008

**Pull request merged**